proposals/generic-ref-struct.md

```diff
@@ -0,0 +1,424 @@
+
+- Instances of `T` cannot be boxed or appear on the heap
+- Instances of `T` can only appear as fields inside a `ref struct`
+- Instances of `T` are considered `ref struct` for the purposes of the [span safety rules](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/span-safety.md) 
+- `T` cannot be used as the element of an array
+- Members on `object` such as `ToString` cannot be invoked as that implicitly boxes
+- etc ...
+
+This also means that `T : where ref struct` cannot be used where `ref struct` would be used in type positions: 
+
+```c#
+class C<T> where T : ref struct 
+{
+    // Error: field cannot be a `ref struct` inside a class
+    T field; 
+
+    // Error: array elements cannot be of type `ref struct`
+    void M(T[] array)
+    {
+
+    }
+}
+```
+
+Positionally in syntax the `ref struct` annotation must come after the type kind, such as `struct`, but before the interface list or `new()` constraint (interfaces are covered [later in the proposal](#interfaces)). The combination with type kinds has the following effects:
+
+- `where T : struct, ref struct`: describes a type parameter that can be a `struct` or `ref struct` but not other types like `class` or `interface`.
+- `where T : unmanaged, ref struct`: similar to `struct, ref struct` except that the type also meets the `unmanaged` constraint. This is believed to have limited use though as `ref struct` containing any type of `ref` data will not meet the `unmanaged` contraint and hence are likely better defined as simply `struct`.
+- `where T : notnull, ref struct`: this is the same as the `notnull` constraint but also allows for `ref struct` types to satisfy it.
+- `where T : class, ref struct`: is an error as the `ref struct` portion is meaningless when combined with `class`. 
+- `where T : default, ref struct`: is an error as the declarations are contradictory.
+- `where T : ref struct, BaseType` where `BaseType` denotes a reference type. This is an error because the `ref sruct` portion is meaningless in this context.
+
+The `where T : ref struct` annotation will be encoded by attributing the type parameter with the attribute `System.Runtime.CompilerServices.SupportsByRefLike`. This will not result in a `modreq` on the containing member nor will `[Obsolete]` tricks be added to the containing type. The intent is for this to be a binary compatible change. The explicit intent is for us to allow types in libraries to relax their type parameters to accept `ref struct` without breaking consumers.
+
+### Supporting Span<T>
+A key motivation of this proposal is to allow for instantiations like `Span<Span<T>>`. Effectively moving the `Span<T>` type to have the `where T : ref struct` annotation. This is a problem because a number of the APIs on `Span<T>` violate the rules of `where T : ref struct` as `T[]` is a common type in the APIs. 
```

Comment:
```suggestion
A key motivation of this proposal is to allow for instantiations like `Span<Span<T>>`, effectively moving the `Span<T>` type to have the `where T : ref struct` annotation. This is a problem because a number of the APIs on `Span<T>` violate the rules of `where T : ref struct` as `T[]` is a common type in the APIs. 
```